src/main/java/org/elasticsearch/index/engine/internal/InternalEngine.java

```diff
@@ -1681,8 +1681,20 @@ public int get() {
 
         public void endRecovery() throws ElasticsearchException {
             store.decRef();
-            onGoingRecoveries.decrementAndGet();
+            int left = onGoingRecoveries.decrementAndGet();
             assert onGoingRecoveries.get() >= 0 : "ongoingRecoveries must be >= 0 but was: " + onGoingRecoveries.get();
+            if (left == 0) {
```

Comment:
this is protected in the flush logic it self, so we can safely try:

https://github.com/elasticsearch/elasticsearch/blob/1.4/src/main/java/org/elasticsearch/index/engine/internal/InternalEngine.java#L841

https://github.com/elasticsearch/elasticsearch/blob/1.4/src/main/java/org/elasticsearch/index/engine/internal/InternalEngine.java#L855

and on the recovery side:

https://github.com/elasticsearch/elasticsearch/blob/1.4/src/main/java/org/elasticsearch/index/engine/internal/InternalEngine.java#L1125